### ISSUE TITLE
strongly typing variable to have set() method

### ISSUE THREAD

**Issue by blubberbo** [szokodiakos/typegoose#264]
I would like to strongly type a variable (in my case a function parameter) based on my model/schema but when I try to, I get an error saying the `.set(` method is undefined.

This is my model:
```
import { RSLTable } from './rsl-table.model';
import { prop, Typegoose, arrayProp } from 'typegoose';
import * as mongoose from 'mongoose';

export class Analyte extends Typegoose {
  @prop()
  chemical: string;
  @prop()
  casrn: string;
  @arrayProp({ items: RSLTable })
  epaRSLTables: RSLTable[];
  @prop({ _id: false })
  epa: RSLTable;
}

export const AnalyteSchema = new Analyte().getModelForClass(Analyte, {
  existingMongoose: mongoose,
  schemaOptions: { collection: 'analytes' },
});
```
and then I have a function such as:
```
export function removeNullValuesFromDoc(doc: Analyte) {
  // check if the doc has an epa value
  if (doc.epa) {
    // the doc has an epa, so check for soil, air and tapWater and remove them if not
    if (!doc.epa.soil) {
      doc.set('epa.soil', undefined);
    }
...
```
but `doc.set(` has red squigglies and says: 
`Property 'set' does not exist on type 'Analyte'.ts(2339)`

Is there a way for me to get around that error so I keep strongly typing things like I like to?

**Comment by julihoh** [comment]
`InstanceType<T>` is what you are looking for.

If you check the return type of `AnalyteSchema.find()` you will note it returns `InstanceType<Analyte>[]` (approximately).

Therefore you should take `InstanceType<Analyte>` as a parameter. Make sure to import `InstanceType` from typegoose as there is a naming conflict with the built in `InstanceType` of TypeScript: 

```typescript
import { InstanceType } from "typegoose";
```

btw: you can also type your `epaRSLTables` better. I use these two small utility types for typing subdocuments:

```typescript
import { Types } from "mongoose";
import { InstanceType, Typegoose } from "typegoose";

/**
 * Specifies that this member is a SubDocument.
 * Note: This is an approximation. Not all members of the resulting type are necessarily available (eg. __v)!
 */
export type SubDocument<T extends Typegoose> = InstanceType<T> & Types.Subdocument;

/**
 * Specifies that this member is an array of SubDocument.
 * See notes for SubDocument for caveats.
 */
export type SubDocumentArray<T extends Typegoose> = Types.DocumentArray<SubDocument<T>>;
```
Your `epaRSLTables` would then be a `SubDocumentArray<RSLTable>`.

**Comment by blubberbo** [comment]
Thank you, the `InstanceType` worked perfectly.

As for the SubDocument, what benefits does this add exactly? Thank you in advance!

**Comment by julihoh** [comment]
Glad I could help :)

The benefits are two separate ones:

First, using `DocumentArray` allows you to use mongoose specific methods on arrays such as: `arr.id(documentId)` which will return the subdocument with the specified document id or (perhaps more importantly) `arr.set(index, val)` which allows you to set the value at a specific index of the array with mongoose's change tracking (meaning mongoose will automatically figure out the correct update query once you save the root document) .

Second, using the `Subdocument` type will allow you to use document specific methods and properties similar to `InstanceType`.
For details see the [mongoose typings](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mongoose/index.d.ts#L1504) on definitelytyped.
Basically it all comes down to typing all mongoose specific methods of these objects.

Hope this explanation helps, let me know if some things are still unclear.

btw: maybe this kind of stuff should be explained in the "Types" section of the readme @maintainers.

**Comment by hasezoey** [comment]
@Ben305 i think this can be closed
